core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultAwsV4HeaderHttpSigner.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.internal;
+
+import static software.amazon.awssdk.http.auth.internal.util.SignerConstant.AWS4_SIGNING_ALGORITHM;
+import static software.amazon.awssdk.http.auth.internal.util.SignerUtils.addChecksumHeader;
+import static software.amazon.awssdk.http.auth.internal.util.SignerUtils.addDateHeader;
+import static software.amazon.awssdk.http.auth.internal.util.SignerUtils.addSha256ContentHeader;
+import static software.amazon.awssdk.http.auth.internal.util.SignerUtils.formatDateTime;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.CompletableFuture;
+import org.reactivestreams.Publisher;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.http.ContentStreamProvider;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.http.auth.AwsV4HttpSigner;
+import software.amazon.awssdk.http.auth.SigV4RequestContext;
+import software.amazon.awssdk.http.auth.internal.checksums.ContentChecksum;
+import software.amazon.awssdk.http.auth.internal.checksums.SdkChecksum;
+import software.amazon.awssdk.http.auth.internal.util.CanonicalRequestV2;
+import software.amazon.awssdk.http.auth.internal.util.SignerConstant;
+import software.amazon.awssdk.http.auth.spi.AsyncSignRequest;
+import software.amazon.awssdk.http.auth.spi.SignRequest;
+import software.amazon.awssdk.http.auth.spi.SyncSignRequest;
+import software.amazon.awssdk.identity.spi.AwsCredentialsIdentity;
+import software.amazon.awssdk.identity.spi.AwsSessionCredentialsIdentity;
+
+/**
+ * An implementation of {@link AwsV4HttpSigner} that adds auth to a request via headers.
+ */
+@SdkProtectedApi
+public final class DefaultAwsV4HeaderHttpSigner implements AwsV4HttpSigner<AwsV4HttpProperties> {
```

Comment:
I will likely change the naming of all of the implementations to remove the `Default`, but couldn't do it while the sub-interfaces exist because of a checkstyle-rule.